Jsonloader

yllekn opened the pull request.

powerjg commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -0,0 +1,55 @@
+        elif d['type'] == 'Vector':
+            return Group(**d)
+    else:
+        return SimStat(**d)
```
Can you set your editor to create newlines at the end of files?

powerjg commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -0,0 +1,55 @@
+        elif d['type'] == 'Group':
+            return Group(**d)
+        elif d['type'] == 'Vector':
+            return Group(**d)
```
Why not 
```suggestion
            return Vector(**d)
```

powerjg commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -0,0 +1,55 @@
+    if 'type' in d:
+        if d['type'] == 'Scalar':
+            return Scalar(
+                          value=d.get('value'),
```
Why `d.get` and not `d['value']`?

yllekn replied:
Some of these statistics have optional data members, and .get() supports returning NoneType (rather than throwing an error). I made all of them .get() for the sake of continuity, but I could differentiate between optional (d.get()) and non-optional data members (d['value']) if you like

powerjg replied:
I see! Would it work to just do the following?

```python
return Scalar(**d)
```

yllekn replied:
I don't think so. These statistics don't take **kwargs as an argument.

yllekn replied:
Let me try it and see what happens though.

powerjg replied:
All python functions take a dictionary as a parameter :smile:

See, for instance, https://book.pythontips.com/en/latest/args_and_kwargs.html

yllekn replied:
I believe you... it works but it requires removing the "type" from the dictionary d, just because of the way we've defined the constructor for statistics and Vectors. I'll push an updated version for you soon to check out

powerjg replied:
Makes sense. One thing we could to is add `**kwargs` as the final parameter to the stats objects to allow other kwargs.

powerjg reviewed (commented):
LGTM! Two small things

powerjg commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -0,0 +1,55 @@
+from .simstat import SimStat
```
Needs a copyright

powerjg commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -0,0 +1,55 @@
+from .group import Group, Vector
+import json
+
+def json_loader(path: str) -> SimStat:
```
maybe just call this `load`? Is there a way it could have the *same* interface as `json.load`?

yllekn replied:
Maybe we don't even need this wrapper function then. The user can just call 
> simstat_obj = json.load(f, object_hook=__json_to_simstat)
like in line 9.

powerjg replied:
I like the idea of hiding the `object_hook` from "regular" users.

yllekn replied:
Ok. I will look into subclassing the JSONDecoder then.

yllekn replied:
Would something like json.loads(f, cls=JSONLoader) be acceptable, where JSONLoader is a subclass of JSONDecoder with object_hooks overwritten? I am trying to think of a way that we could hide the underlying "JSONLoader" stuff from user entirely

yllekn commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -1,55 +1,72 @@
+    Subclass of JSONDecoder that overrides 'object_hook'. Converts JSON object
+    into a SimStat object.
+
+    Usage
```
@powerjg What do you think of this interface?

yllekn commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -1,55 +1,72 @@
+    def __json_to_simstat(self, d: dict) -> SimStat:
+        if 'type' in d:
+            if d['type'] == 'Scalar':
+                d.pop('type', None)
```
Not ideal...

powerjg commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -0,0 +1,72 @@
+from .group import Group, Vector
+import json
+
+class JSONLoader(json.JSONDecoder):
```
This is *perfect*!

I suggest also adding a function `load` so you can use the following:
```
with open('stats.json') as f:
  pystats.jsonloader.load(f)
```
This load function should be *similar* (though not the same) as the `json.load` function.

yllekn replied:
So just like a wrapper function for json.load?

powerjg replied:
Yeah, exactly!

powerjg commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -0,0 +1,72 @@
+    def __json_to_simstat(self, d: dict) -> SimStat:
+        if 'type' in d:
+            if d['type'] == 'Scalar':
+                d.pop('type', None)
```
I suggest dropping these. Instead, I think that `Scalar`, `Vector`, etc. should be modified something like below...

```
class Scalar(Statistic):
    """
    A scalar Python statistic type.
    """

    value: Union[float, int]

    def __init__(self, value: Any,
                 unit: Optional[str] = None,
                 description: Optional[str] = None,
                 datatype: Optional[StorageType] = None,
                 **kwargs = None):
        super(Scalar, self).__init__(
                                     value=value,
                                     type="Scalar",
                                     unit=unit,
                                     description=description,
                                     datatype=datatype,
                                     **kwargs
                                    )
```

yllekn replied:
Does this solve the problem? (I will try this in just a second to find out.) I don't think the issue was that the constructor wouldn't take additional arguments, but it was because it was receiving two of the same argument, e.g. 'type' was received twice?

powerjg replied:
I *think* so. I meant to tag @BobbyRBruce to see what he thinks about this as well. 

I could be wrong, though

yllekn replied:
I tried it, and ran into the same error from before I added the d.pop() statements:

`  File "build/NULL/python/m5/ext/pystats/jsonloader.py", line 53, in __json_to_simstat
  File "build/NULL/python/m5/ext/pystats/statistic.py", line 72, in __init__
TypeError: __init__() got multiple values for keyword argument 'type'`

powerjg replied:
Gotcha... Oh! I see... 
```
( value=value, type="Scalar",
...
```

I'm good with this change as it is, then :)

yllekn commented on src/python/m5/ext/pystats/jsonloader.py:
````diff
@@ -1,55 +1,72 @@
\ No newline at end of file
+    simstat_object = json.load(f, cls=JSONLoader)
+    ```
+    """
````
@powerjg What do you think of this interface, where caller must specify 'cls' argument?

yllekn replied:
Sorry, not sure why this comment is so weirdly formatted. I'm referring to the 'Usage'

yllekn commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -1,55 +1,72 @@
+                return Group(**d)
+
+            elif d['type'] == 'Vector':
+                d.pop('type', None)
+                d.pop('time_conversion', None)
```
d.pop() is not ideal, but it works

powerjg replied:
@BobbyRBruce, why does vector not have a time conversion, but everything else does...?

BobbyRBruce replied:
It does have time_conversation. Vector inherets from "Group" which contains it.

powerjg replied:
So, why `pop('time_conversion')`, @yllekn?

yllekn replied:
The constructor for Vector has 'time_conversion = None' in it. I'm not sure what the intent was, or if it was meant to be an optional argument, but it's the same issue as before with 'type'-- setting the property 'time_conversion' twice.

yllekn replied:
In other words, I get that same 'multiple values' error.

powerjg reviewed (commented):
Looks great!

yllekn commented on src/python/m5/ext/pystats/jsonloader.py:
````diff
@@ -70,3 +70,20 @@ def __json_to_simstat(self, d: dict) -> SimStat:
         else:
             return SimStat(**d)
+
+def load(json_file) -> SimStat:
+    """
+    Takes a JSON file translates it into a SimStat object.
+
+    Usage
+    -----
+    ```
+    import m5.ext.pystats as pystats
+
+    with open(path) as f:
+        pystats.jsonloader.load(f)
+    ```
+    """
+
+    simstat_object = json.load(json_file, cls=JSONLoader)
+    return simstat_object
````
@powerjg

powerjg commented:
Did you do a force push to the jsonloader branch?

@BobbyRBruce, could you do a quick review before we push to gerrit?

yllekn commented:
@powerjg Sorry I forgot that you mentioned that. It's done now.

powerjg commented on src/python/m5/ext/pystats/simstat.py:
```diff
@@ -42,10 +42,10 @@ class SimStat(JsonSerializable):
-                 time_conversion: Optional[TimeConversion],
-                 simulated_begin_time: Optional[Union[int, float]],
-                 simulated_end_time: Optional[Union[int, float]],
+    def __init__(self, creation_time: Optional[datetime] = None,
```
This should be a separate commit:
"python: Fix optional parameters with default None"

BobbyRBruce replied:
Agreed.

yllekn replied:
I can try to figure out how to do this

yllekn replied:
resolved [here](https://github.com/darchr/gem5/compare/d974742952a4acfd9a5711f35d25cd42e84d0e70..bf98acaa9d8631db3a97f15eb4bd06524d690d68)

powerjg commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -0,0 +1,89 @@
+# Copyright (c) 2020 The Regents of The University of California
```
```suggestion
# Copyright (c) 2021 The Regents of The University of California
```

yllekn replied:
lol

yllekn replied:
resolved [here](https://github.com/darchr/gem5/compare/d974742952a4acfd9a5711f35d25cd42e84d0e70..bf98acaa9d8631db3a97f15eb4bd06524d690d68)

powerjg reviewed (commented):
This is good with me. I'd like @BobbyRBruce to take a quick look, too.

My comments are now about how best to push to gerrit, not on the changes themselves ;).

BobbyRBruce commented on src/python/m5/ext/pystats/__init__.py:
```diff
@@ -30,6 +30,7 @@
 from .statistic import Statistic
 from .storagetype import StorageType
 from .timeconversion import TimeConversion
+from .jsonloader import JSONLoader
```
It's not a big problem but I do format this as "Json" for other classes. E.g., "JsonSerializable"

yllekn replied:
resolved [here](https://github.com/darchr/gem5/compare/d974742952a4acfd9a5711f35d25cd42e84d0e70..bf98acaa9d8631db3a97f15eb4bd06524d690d68)

BobbyRBruce commented on src/python/m5/stats/gem5stats.py:
```diff
@@ -230,7 +230,7 @@ def _prepare_stats(group: _m5.stats.Group):
         stat.prepare()
 
-    for child in getStatGroups().values():
+    for child in group.getStatGroups().values():
```
Just be careful here. This has already been fixed: https://gem5-review.googlesource.com/c/public/gem5/+/42704 (this will be merged into develop this afternoon: https://gem5-review.googlesource.com/c/public/gem5/+/43265).

yllekn replied:
Hm. How to deal with this then? Wait until this is merged with develop and then rebase or something?

powerjg replied:
Exactly :)

yllekn replied:
@powerjg Doesn't look like this change is reflected in the github repo, am I supposed to cherry-pick this commit from googlesource or something like that?

BobbyRBruce commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -0,0 +1,89 @@
+                d.pop('time_conversion', None)
+                return Vector(d)
+        else:
+            return SimStat(**d)
```
Can we have some fail-cases in `__json_to_simstat` function? I'm thinking if someone adds a new stat type, this function should throw an "unrecognized type" error.

I'm also a bit uncomfortable that the definition of a SimStat is anything that doesn't have a "type". Is there a way we can improve this somehow? Maybe give SimStat a type of "simstat" ? It would be worrying if someone extended the Python Stats with something that doesn't have a type and they are cast to SimStats (I guess this would throw an error somewhere but it might be hard to diagnose?).

yllekn replied:
Maybe @powerjg could weigh in on this

powerjg replied:
I think this is fine for now. I don't know how better to do this if we're deserializing from json. I think we're going to want to provide simplified output formats (i.e., without the type) and then we'll have to make even more "guesses" with the deserializer.

If you know better ways to solve this problem, Bobby, I'm all ears.

BobbyRBruce commented on src/python/m5/ext/pystats/jsonloader.py:
```diff
@@ -0,0 +1,89 @@
+    from m5.ext.pystats.jsonloader import JSONLoader
+
+    with open(path) as f:
+        simstat_object = json.load(f, cls=JSONLoader)
```
I got a little confused here due to the function at https://github.com/darchr/gem5/pull/218/files#diff-7d72dc1dd99fe1127f61d32392d0943663f4187f7249fcfd68c667bdae10530fR74 . Perhaps it's best to mention here that a helper function is provided to use this class in a cleaner way.

yllekn replied:
resolved [here](https://github.com/darchr/gem5/compare/d974742952a4acfd9a5711f35d25cd42e84d0e70..bf98acaa9d8631db3a97f15eb4bd06524d690d68)

yllekn replied:
resolved [here](https://github.com/darchr/gem5/compare/d974742952a4acfd9a5711f35d25cd42e84d0e70..bf98acaa9d8631db3a97f15eb4bd06524d690d68)

BobbyRBruce reviewed (commented):
Well done Kelly! Though i do have some comments, I'm mostly happy with these changes. I'm surprised how clearly the Hooks works. If there are next steps to this I'd recommend:

- Tests which run in our TestLib Framework. I think, as a minimum, check that the SimStats from a basic simulation can be translated to and from JSON while preserving all the data (I would actually run a small simulation for this so we know the PyBindings are still working and the translation is still good). 

- Someone needs to update Jason's schema: https://github.com/gem5/stats-schema. It's out of date due to some of the latest changes.

yllekn committed:
python: Add JsonLoader to pystats

Allows for loading JSON files (e.g. m5out/stats.json) into the gem5 Python
stats model.

Change-Id: I948dff0b23ea65c4c7acaac5e4fdf3795de1c0a7
Signed-off-by: Kelly Nguyen <klynguyen@ucdavis.edu>

yllekn committed:
python: Fix optional parameters with default None

Change-Id: I40c80c8fdd1384b3405e6b1c0b0b9f310580568d
Signed-off-by: Kelly Nguyen <klynguyen@ucdavis.edu>

yllekn commented:
I rebased the branch to split a commit for Optional parameters and make some other changes. I'll get to testing per Bobby's suggestion soon. Would appreciate some feedback on the above unresolved issues @powerjg @BobbyRBruce

powerjg reviewed (commented):
I think one more rebase is in order, but then you can push to gerrit. I think there should be two commits: 1. fix the missing nones 2. Add the json loader.
